rfc/rfc-64/rfc-64.md

```diff
@@ -0,0 +1,509 @@
+    INCREMENTAL,
+    // Representing state of the table as a stream of CRUD operations on
+    // individual records
+    CDC
+}
+```
+
+**Expression** (**Predicate**)
+
+We will need to introduce our own Expression hierarchy supporting a *superset* of the 
+commonly used expressions for the following reasons:
+1. We need to do deep introspections into these expression trees (in the analyses we
+   run w/in Column Stats Index for ex)
+2. However, we won't be implementing an *Expression Execution Engine*: for execution
+   we will continue to rely on query engine's machinery simply relaying expression-trees from
+   our own representation into the engine-specific ones
+
+```java
+interface Expression {
+  Expression[] children();
+}
+
+// Unary expressions like NOT, CAST, common single argument functions like sin, abs, etc,
+interface UnaryExpression {
+  Expression child();
+}
+
+// Binary expressions like AND, OR, GT, etc
+interface BinaryExpression {
+  Expresion left();
+  Expression right();
+}
+
+// TODO clarify if we need this delineation 
+// Simplified representation of the more expansive `Expression` tree structure (for ex, 
+// used for filter push-down)
+class Predicate {
+  // ...
+}
+```
+
+#### Mid-level (engine-friendly) APIs
+
+**TimelineView** interface will be providing lightweight view of Hudi's timeline to be
+able to correlate timestamps to w/ actions on Hudi's timeline
+
+```java
+interface TimelineView {
+    // Looks up latest commit instant, provide (temporal) instant T;
+    // This method is required to resolve commit-instants by the timestamp
+    HoodieInstant findLatestCompletedActionAt(Instant instant);
+    
+    // ...
+}
+```
+
+##### File Listing
+
+To provide for *listing* capability described above we will be introducing following facades:
+ 
+ - **RecordIndex**: to be able to resolve records against the file-groups where such records
+are persisted (*required for writing*)
+ - **FileIndex**: to be able to list file(-slices) constituting *partition snapshot*, satisfying 
+pushed-down data-filters (validated t/h Partition Pruning, Files' Columns Stats filtering, etc)
+
+```java
+class RecordIndex {
+    // Tags provided record-keys (PKs) within corresponding locations within the dataset;
+    // in case record-key is not currently present in the dataset, no location will be provided
+    HoodiePairData<HoodieKey, Option<HoodieRecordLocation>> tag(HoodieData<HoodieKey> records);
+}
+
+class FileIndex {
+    // Lists files visible/reachable at the instant T
+    // Equivalent to `listFilesBetween(null, instant, filters)`
+    List<PartitionSnapshot> listFilesAt(HoodieInstant instant, Predicate[] filters);
+    // Lists files added visible/reachable at the instant `to`, that were
+    // added no earlier than at the instant `from`
+    List<PartitionIncrementalSnapshot> listFilesBetween(HoodieInstant from, HoodieInstant to, Predicate[] filters);
```

Comment:
will FileIndex cache all file information to memory just like BaseHoodieTableFileIndex？
if yes， I don't think it's friendly to presto

Reply:
@xiarixiaoyao this will be configurable